### ISSUE TITLE
Create a script to remove empty headings from CMS Content

### ISSUE THREAD

**Issue by bibliotechy** [eScholarship/jschol#995]
Several accessibility issues raised by the UX team were focused on CMS content. It appears there is no way for the CMS editors to remove this so we need a script to remove these from existing content.

**Comment by DevinSmithWork** [comment]
- [x] Assess where CMS content is stored (pairtree? DB? other?)
- [x] Pull down some examples
- [x] Create py ET parsing script
- [x] Gameplan deployment

**Comment by DevinSmithWork** [comment]
Examples in parent ticket: https://github.com/orgs/eScholarship/projects/1/views/9?pane=issue&itemId=160166967&issue=eScholarship%7Cjschol%7C964

**Comment by DevinSmithWork** [comment]
Pages table includes the data for remediation:
```SQL
select
	p.*,
	p.attrs ->> "$.html" as html
from pages p
where p.slug = 'publishing'
```

Example HTML with errors:
```HTML
<h2><strong></strong></h2>

<h1>What is eScholarship Publishing?</h1>

<p>Established in 2002, eScholarship Publishing is a comprehensive open access publishing program for the University of California academic community. eScholarship offers publishing and production tools as well as professional support and consulting services for journals, preprints, monographs, conference proceedings, and other UC-affiliated original scholarship. Our publications often traverse standard disciplinary boundaries, explore new publishing models, support under-represented voices within the scholarly record, and reach communities and/or professionals in applied fields beyond academia.</p>

<h1><strong></strong></h1>

<h1>Why should I publish with eScholarship?</h1>

<p>eScholarship is committed to open access publishing, ensuring global access to research publications and emphasizing the rights of authors to retain copyright in their work and to license that work in ways that encourage sharing and reuse.</p>

<p>eScholarship also offers publications increased visibility, speed to publication, and a commitment to long-term access and 
<a href="https://merritt.cdlib.org/">preservation</a>.</p>

<h1>Who is behind eScholarship Publishing?</h1>

<h1></h1>

<p>eScholarship Publishing is a library publishing program managed by the <a href="https://cdlib.org/">California Digital Library
</a>, supported by the University of California and offered free of charge to UC-affiliated departments, research units, publishing programs, and individual scholars.<br></p>

<p></p>

<hr>

<p></p>

````

**Comment by DevinSmithWork** [comment]
### Findings:

[prod-pages-with-empty-elements.tsv](https://github.com/user-attachments/files/26450193/prod-pages-with-empty-elements.tsv)

- ~900 pages (out of ~4,500) on eSchol have empty elements (see attached)
- Most are empty `p`s, a few involve odd/wrong nesting, and there's a few extremely weird (ambitious?) outliers
- Script to find and generate remediations is in place, but @bibliotechy we should probably gameplan whether or how much human intervention we would like before performing bulk-updates the table.
   - E.g., as a first pass, eyeballing the empty elements list on the doc and flagging anything other than garden-variety `p`s `h`s, etc?
   - Or, shall we start with only `/root` pages and liaise w/ journal editors to confirm remediations?

**Comment by DevinSmithWork** [comment]
Sprint meeting: Awaiting Chad's return to health, Devin will book a meeting w/ Justin & journals team to assess Qs regarding human-in-the-loop QA for remediated HTML.

**Comment by DevinSmithWork** [comment]
## Human-in-the-loop meeting 2026-04-09: Takeaways

- [x] DS: Doublecheck text encoding (etc) for weird handling of HTML-specific escape seqs like `&nbsp;`
   - [ ] If convenient, loop in spreadsheet boolean for orphan pages (see Alainna DM from 2026-04-09)
- [x] DS: Identify "low-hanging-fruit" pages w/ only one-or-two empty `p` or `a` elements
- [x] DS: Implement demo remediation on stg (idea: Create new pages w/ stub `page_name_rem` for easy 1:1 viewing?)
- [x] Pass spreadsheet w/ links to demo stg. remediation pages to journals team (Justin, Alaina, Charlotte, Amanda)

### Steps following:
- Low-hanging fruit: Go ahead and update the DB except for `root` and top-level campus pages (e.g. `ucsb/about`)
   - Coordinate in-person review of these pages w/ journals team
- Assess non-low-hanging-fruit before/after when we get there.

**Comment by DevinSmithWork** [comment]
@alainna @justingonder 

This sheet contains links to demo remediations on staging for easy visual comparison. The last two columns "STG URL" and "REM STG URL" are before/after. I somewhat randomly selected around 40 from the ~600 pages on Staging with empty elements. 

Please take a gander and LMK if anything looks broken.

[remediation_demos_on_staging.xlsx](https://github.com/user-attachments/files/26686273/remediation_demos_on_staging.xlsx)

I've already spotted one thing that need fixing: Images are not always parsed as expected (e.g. the journal logo is missing in `cfp_rem`) — this means ~it's possible~ other self-closing tags (e.g., `<br/>` and `<hr/>`) ~might~ have similar issues -- DS: Confirmed, they do

**Comment by DevinSmithWork** [comment]
### Low-hanging fruit

1–2 elements is about half.

| elem | page count | RT |
| -: | -: | -: |
| 1 | 181 | 181 |
| 2 | 269 | 450 |
| 3 | 24 |  475 |
| 4 | 100 | 575 | 
| 5 | 21 | 596 |
| 6+ | 71 | 667|

**Comment by DevinSmithWork** [comment]
Google doc w/ Spreadsheet, including pubs for manual review: https://docs.google.com/spreadsheets/d/1T6FFLyrQJoeKgKmY8yGKMm2qLMzti_woK8_I2nxvrjE/edit?gid=0#gid=0

**Comment by DevinSmithWork** [comment]
- Empty element removal almost completed.
- A small handful of egregiously bad (non-empty-element) HTML remediations to follow (e.g., DOJ editor pages).
- During this project, I've found several instances of what I believe is our WYSIWYG editor inserting empty elements in certain circumstances (e.g., surrounding `<hr>` and possible `<ol>` elements). Unsure about the priority of additional work on this. Discuss

**Comment by DevinSmithWork** [comment]
- Empty element remediation complete. 
   - DOJ editorial pages reformatted.
   - A few manual HTML fixes for odd circumstances.
- Remediation program uploaded to github: https://github.com/eScholarship/eschol-pages-html-remediation

### Notes:
- eSchol pages still include a fair amount of HTML that isn't great for accessibility, but is much more difficult/impossible to remediate automatically.
   - The most common problem is people using `<p><strong>Text</p></strong>` for headings, where they should be using `<h>` elements.
   - Material that's been copied/pasted from word processors (etc) sometimes don't resolve bulleted or numbered lists into `<ul>` or `<ol>` (respectively), instead nesting the list indicating as text, e.g.: `<p>• List item</p>`
   - Stacking `<a>` elements on text with `<strong>` or `<em>` appears to sometimes generate additional `<a>` elements with no text.

### Going Forward
- Monthly crontab the remediation program on `pub-oapi-tools` ec2.
- Assess:
   - Priority of lower-priority HTML remediations? (e.g. `<p><strong>` to `<h>`)
   - Modifying or replacing the WYSIWYG editor?

**Comment by justingonder** [comment]
Re-run monthly with report-out
